### PR TITLE
[TRTLLM-15760][refactor] Establish model runner contracts and shared model invocation

### PR DESCRIPTION
**Pull request by lori-ren**
## Dev Engineer Review
- The change establishes separate scheduled and packed runner contracts. `ScheduledInputs` and `PackedInputs` carry per-call options into runner execution.
- Runners now own input buffers and receive mapping, distribution, load-balancer, and model-caller dependencies directly. `ModelCaller` centralizes model invocation and compilation context.
- `PyTorchModelEngine` retains an engine-owned decoder fallback. Decoder state and resource migration remain outside this change.
- `ModelEngine.forward` and `ADEngine.forward` no longer accept `gather_context_logits` as a separate argument. Callers must provide the option through packed inputs or context-request metadata.
- Review severity counts are unavailable because no current review findings were supplied.

## QA Engineer Review
- Modified tests cover runner contracts, model invocation, resource validation, warmup and graph lifecycle, CUDA-graph replay, encoder forwarding, and context-logit handling.
- The supplied changes and test-list search show no test-list edits or matching CI/manual-QA entries for these unit tests. An existing waiver references `test_pytorch_model_engine.py`; this is not evidence that the test is listed for execution.
- Test execution results were not supplied. Coverage verdict: **needs follow-up**.

## Per-File QA Perspective
- `tensorrt_llm/_torch/auto_deploy/shim/ad_executor.py` — Verify that context requests set `py_return_context_logits` as expected and that speculative-decoding validation receives the derived value.
- `tensorrt_llm/_torch/pyexecutor/encoder_executor.py` — Verify encoder calls pass `PackedInputs` and preserve the context-logit option.
- `tensorrt_llm/_torch/pyexecutor/engine/input_buffers.py` — Verify allocations for cache-indirection settings, beam widths, and draft-token capacities, including `None` and zero.
- `tensorrt_llm/_torch/pyexecutor/engine/lora.py` — Verify constructor-injected CUDA-graph LoRA manager behavior and eager fallback when it is absent.
- `tensorrt_llm/_torch/pyexecutor/engine/model_call.py` — Verify tracing, compile context, metadata propagation, and behavior when the model raises an exception.
- `tensorrt_llm/_torch/pyexecutor/engine/runners/__init__.py` — Verify runtime runner-type resolution after the return-type update.
- `tensorrt_llm/_torch/pyexecutor/engine/runners/encoder.py` — Verify runner-owned buffer transfers, attention metadata, context-logit gathering, and graph release.
- `tensorrt_llm/_torch/pyexecutor/engine/runners/encoder_decoder.py` — Verify the scheduled-input contract and resource-manager forwarding during warmup and execution.
- `tensorrt_llm/_torch/pyexecutor/engine/runners/interface.py` — Verify callers use the new input records and lifecycle hooks after removal of `RunnerDeps` and protocol methods.
- `tensorrt_llm/_torch/pyexecutor/engine/runners/no_kv_cache.py` — Verify resource validation and runner-owned buffer use during warmup and forward execution.
- `tensorrt_llm/_torch/pyexecutor/engine/runners/pooling.py` — Verify pooling uses `ModelCaller` and preserves context-logit behavior.
- `tensorrt_llm/_torch/pyexecutor/model_engine.py` — Verify packed and scheduled dispatch, decoder fallback, output handling, graph lifecycle, and input-copy waits.
- `tensorrt_llm/_torch/pyexecutor/py_executor.py` — Verify context-logit behavior now that `PyExecutor` no longer passes the option to `ModelEngine.forward`.
- `tests/unittest/_torch/executor/engine/test_encoder.py` — Covers packed inputs, attention metadata setup, and graph release. No matching CI or manual-QA list entry was found.
- `tests/unittest/_torch/executor/engine/test_encoder_decoder.py` — Covers scheduled-input and preparation contracts. No matching CI or manual-QA list entry was found.
- `tests/unittest/_torch/executor/engine/test_lora.py` — Covers constructor-injected CUDA-graph LoRA behavior and eager fallback. No matching CI or manual-QA list entry was found.
- `tests/unittest/_torch/executor/engine/test_no_kv_cache.py` — Covers no-KV-cache runner setup and input preparation. No matching CI or manual-QA list entry was found.
- `tests/unittest/_torch/executor/engine/test_runners.py` — Covers dependency wiring, resource rejection, model-caller context, output handling, and lifecycle hooks. No matching CI or manual-QA list entry was found.
- `tests/unittest/_torch/executor/test_eager_workspace_engine.py` — Covers model-caller test wiring for eager workspace behavior. No matching CI or manual-QA list entry was found.
- `tests/unittest/_torch/executor/test_encoder_executor.py` — Covers encoder executor forwarding of `PackedInputs` and context-logit settings. No matching CI or manual-QA list entry was found.
- `tests/unittest/_torch/executor/test_pytorch_model_engine.py` — Covers engine forwarding, LoRA selection, and CUDA-graph replay with stable buffer pointers. No matching CI or manual-QA list entry was found; an existing waiver references this test file.
- `tests/unittest/_torch/executor/test_pytorch_model_engine_warmup.py` — Covers runner warmup and fallback execution. No matching CI or manual-QA list entry was found.
- `tests/unittest/auto_deploy/singlegpu/shim/test_engine.py` — Covers chunked-prefill equivalence with context logits enabled and disabled. No matching CI or manual-QA list entry was found.
- `tests/unittest/_torch/executor/test_breakable_cuda_graph.py` — Covers CUDA-graph execution through the decoder callback and `PyTorchModelEngine.forward`. No matching CI or manual-QA list entry was found.
- `tests/unittest/_torch/executor/test_py_executor.py` — Covers context-logit collection and `PyExecutor._forward_step` forwarding. No matching CI or manual-QA list entry was found.

## Description

Follow-up to #18933, which introduced a separate encoder execution path without a shared runner abstraction. This PR establishes common contracts for the encoder and existing decoder paths, laying the groundwork for the decoder runner refactor.

Establish explicit contracts for scheduled and packed model runners so decoder execution can move out of ModelEngine incrementally. Batch inputs carry per-call options, runners own their input buffers, and shared model invocation preserves execution metadata and compilation context.

Adapt encoder, pooling, multimodal, and engine callers to these contracts while retaining the engine-owned decoder fallback. Decoder state and resource migration remain follow-up work.

## Test Coverage

- `tests/unittest/_torch/executor/engine/`
- `tests/unittest/_torch/executor/test_pytorch_model_engine.py`
- `tests/unittest/_torch/executor/test_pytorch_model_engine_warmup.py`
- `tests/unittest/auto_deploy/singlegpu/shim/test_engine.py`

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 1]
```diff
@@ -6113,54 +6117,81 @@ def _prepare_inputs(
...
             )
+
+        # Transitional: warmup and memory profiling both set engine.is_warmup.
+        # Decoder helpers still read that field for dummy-pass behavior; the
+        # runner flag does not establish the warmup scope or its MoE effects.
+        assert is_dummy == self.is_warmup
+        self.enable_spec_decode = inputs.enable_spec_decode
+        self.runtime_draft_len = inputs.runtime_draft_len
+        outputs = self._forward_decoder(inputs, resource_manager)
+        return {**outputs, "runtime_draft_len": self.runtime_draft_len}
```
[the remark this answers is not in the document]
Comment (lori-ren) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:6184]: Guarded both decoder output unpacking and draft-length extraction. Raw-output and BCG-through-engine regression tests pass.

Fixed in 3679b748b7.

File: tests/unittest/_torch/executor/engine/test_runners.py [hunk 2]
```diff
@@ -603,3 +616,93 @@ def test_mm_encoder_runner_splits_embeddings_and_returns_mrope_metadata() -> Non
...
+def test_engine_consumes_optional_length_update_and_passes_call_state(effective_length, is_dummy):
+    runner = Mock(spec=NoKVCacheRunner)
+    outputs = {"logits": object()}
+    if effective_length is not None:
+        outputs["runtime_draft_len"] = effective_length
+    runner.forward.return_value = outputs
+    engine, resources = _model_engine_with_runner(runner, kv_cache_manager=None)
+    engine.runtime_draft_len = 2
+    engine.enable_spec_decode = True
+    engine.is_warmup = is_dummy
```
[the remark this answers is not in the document]
Comment (lori-ren) [reply tests/unittest/_torch/executor/engine/test_runners.py]: Set the backing warmup field directly so this test no longer modifies process-wide MoE state.

Fixed in 3679b748b7.

File: tensorrt_llm/_torch/pyexecutor/engine/model_call.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        attrs["attention_metadata"] = weakref.ref(kwargs["attn_metadata"])
+        attrs.update(self._model.model_config.extra_attrs)
+        attrs["spec_metadata"] = kwargs.get("spec_metadata", None)
+
+        if self._compile_backend is not None:
+            # Compilation can update the stream and event containers in place.
+            assert self._aux_streams is not None
+            attrs["aux_streams"] = weakref.ref(self._aux_streams)
+            attrs["events"] = weakref.ref(self._compile_backend.events)
+            attrs["global_stream"] = torch.cuda.current_stream()
```
[the remark this answers is not in the document]
Comment (lori-ren) [reply tensorrt_llm/_torch/pyexecutor/engine/model_call.py:59]: Added compile-stream/event weak-reference tests, including live container changes and missing-stream validation.

Fixed in 3679b748b7.

File: tensorrt_llm/_torch/pyexecutor/model_engine.py [hunk 4]
```diff
@@ -6113,54 +6117,81 @@ def _prepare_inputs(
...
-            gather_context_logits=gather_context_logits,
             cache_indirection_buffer=cache_indirection_buffer,
             num_accepted_tokens_device=num_accepted_tokens_device,
-            req_id_to_old_request=req_id_to_old_request,
+            previous_request_slots=({
+                req_id: request.py_seq_slot
+                for req_id, request in req_id_to_old_request.items()
+            } if req_id_to_old_request is not None else None),
+            gather_context_logits=any(request.py_return_context_logits
+                                      for request in batch.context_requests),
```
[the remark this answers is not in the document]
Comment (lori-ren) [reply tensorrt_llm/_torch/pyexecutor/model_engine.py:6142]: Added coverage through engine.forward for previous-request slot derivation and forwarding to the decoder.

Fixed in 3679b748b7.

File: tensorrt_llm/_torch/pyexecutor/py_executor.py [hunk 5]
```diff
@@ -7545,20 +7545,15 @@ def _forward_step(
             f"[Executor] _forward_step {self.iter_counter}: {scheduled_requests.num_context_requests} ctx reqs, {num_ctx_tokens} ctx tokens, {scheduled_requests.num_generation_requests} gen reqs"
         )
         def forward(scheduled_requests, resource_manager, new_tensors_device,
-                    gather_context_logits, cache_indirection_buffer,
-                    num_accepted_tokens_device):
+                    cache_indirection_buffer, num_accepted_tokens_device):
```
[the remark this answers is not in the document]
Comment (lori-ren) [reply tensorrt_llm/_torch/pyexecutor/py_executor.py:7548]: Added executor-to-engine-to-runner coverage for batch-carried context-logit requests.

Fixed in 3679b748b7.

File: tests/unittest/auto_deploy/singlegpu/shim/test_engine.py [hunk 6]
```diff
@@ -287,6 +287,7 @@ def __init__(self, tokens: List[int], begin: int, size: int, seq_slot: int = 0):
         self.py_seq_slot = seq_slot
         self.py_request_id = seq_slot
         self.py_batch_idx = None
+        self.py_return_context_logits = False
```
[the remark this answers is not in the document]
Comment (lori-ren) [reply tests/unittest/auto_deploy/singlegpu/shim/test_engine.py:290]: Added context-logit coverage for full and chunked prefill while retaining ordinary-logit cases.

Fixed in 3679b748b7.
